torchaudio_contrib/augmentation.py

```diff
@@ -0,0 +1,151 @@
+import math
+import torch
+
+import torch.nn as nn
+from .layers import _ModuleNoStateBuffers, STFT
+from .functional import phase_vocoder
+
+
+class StretchSpecTime(_ModuleNoStateBuffers):
+    """
+    Stretch stft in time without modifying pitch for a given rate.
+
+    Args:
+
+        rate (float): rate to speed up or slow down by. Defaults to 1.
+        hop_len (int): Number audio of frames between STFT columns.
+            Defaults to 512.
+        num_bins (int, optional): number of filter banks from stft.
+            Defaults to 1025.
+    """
+
+    def __init__(self, rate=1., hop_len=512, num_bins=1025):
+
+        super(StretchSpecTime, self).__init__()
+
+        self.rate = rate
+        phi_advance = torch.linspace(
+            0, math.pi * hop_len, num_bins)[..., None]
+
+        self.register_buffer('phi_advance', phi_advance)
+
+    def forward(self, spect, rate=None):
+        if rate is None:
+            rate = self.rate
+        return phase_vocoder(spect, rate, self.phi_advance)
+
+    def __repr__(self):
+        param_str = '(rate={})'.format(self.rate)
+        return self.__class__.__name__ + param_str
+
+
+class GaussianNoise(_ModuleNoStateBuffers):
+    """
+    Apply gaussian noise to a spectrogram.
+
+    Args:
+        scale (float or tensor): standard deviation of the distribution.
+    """
+
+    def __init__(self, scale=1.):
+        super(GaussianNoise, self).__init__()
+        self.scale = torch.as_tensor(scale)
+        self.loc = torch.tensor(0.)
+
+    def forward(self, spect, scale=None):
+        if scale == None:
+            scale = self.scale
+        else:
+            scale = torch.as_tensor(scale)
+
+        with torch.no_grad():
+
+            noise = torch.normal(self.loc.expand(spect.shape), scale.expand(spect.shape)) 
+            return spect, noise
+
+    def __repr__(self):
+        param_str = '(scale={})'.format(self.scale)
+        return self.__class__.__name__ + param_str
+
+
+def apply_masking(spect, max_value, mask_value, axis):
```

Comment:
the name seems slight too general. probably something like `mask_random_region_along_axis`..but shorter? :D

Reply:
also i think we need little discussion at some point about if this masking would be always with zeros (which i think people would assume) or with a number as implemented here. it is more generalizable here, but because if the use-case is always with zeros, i'd expect after the masking the masked region is not valid values with zeros, but just doesn't exist at all (i.e., the output is smaller in size) so that computation could be more efficient.

could be tricky for us, maybe we also need to invent a slightly different name, because for `SpecAug`, actually, the current implementation can make more sense. overall, i'm not 100% sure about this but let's think about it.

Reply:
>also i think we need little discussion at some point about if this masking would be always with zeros (which i think people would assume) or with a number as implemented here.

So the functional doesn't have a default but in the layer I've set the default to 0 (as users would expect) while also allowing to set a desired value in case the spectrogram isn't whitened.

> masking the masked region is not valid values with zeros, but just doesn't exist at all

Oh right... Yeah I this seems different than `SpecAug` but could be done with something like `torch.masked_select`? Although this would have to be used very carefully when fed into a network... (keeping track of lengths and heights, padding, etc.) but cool.

Reply:
> (keeping track of lengths and heights, padding, etc.)

Ugh, I missed that. That would be very very ugly, your implementation should be the way to go IMO.